Objects/exceptions.c

```diff
@@ -194,25 +194,34 @@ _PyBaseExceptionObject_cast(PyObject *exc)
 }
 
 static PyObject *
-BaseException_add_note(PyObject *self_, PyObject *args, PyObject *kwds)
+BaseException_add_note(PyObject *self, PyObject *args, PyObject *kwds)
 {
-    PyBaseExceptionObject *self = _PyBaseExceptionObject_cast(self_);
     PyObject *note = NULL;
 
     if (!PyArg_ParseTuple(args, "U:add_note", &note)) {
         return NULL;
     }
 
-    if (self->notes == NULL) {
-        self->notes = PyList_New(0);
-        if (self->notes == NULL) {
+    if (!PyObject_HasAttr(self, &_Py_ID(__notes__))) {
+        PyObject *new_notes = PyList_New(0);
+        if (new_notes == NULL) {
             return NULL;
         }
+        if (PyObject_SetAttr(self, &_Py_ID(__notes__), new_notes) < 0) {
+            Py_DECREF(new_notes);
+            return NULL;
+        }
+        Py_DECREF(new_notes);
+    }
+    PyObject *notes = PyObject_GetAttr(self, &_Py_ID(__notes__));
+    if (notes == NULL) {
+        return NULL;
     }
-    assert(PyUnicode_CheckExact(note)); 
-    if (PyList_Append(self->notes, note) < 0) {
+    if (PyList_Append(notes, note) < 0) {
+        Py_DECREF(notes);
```

Comment:
This probably needs to be slightly nicer -- if `__notes__` isn't a list this raises `SystemError("bad argument to internal function")`. (Unless there's no way to assign to it? But there's a way to delete it. Somehow I can't find the descriptor/property that gives access to it from user code.)

Reply:
You're right, I will add a check that it's a list. 

This version implement's [the SC's suggestion](https://discuss.python.org/t/pep-678-enriching-exceptions-with-notes/13374/69) where there is no descriptor, so we only allocate space for ``__notes__`` if you use them.   For the typical user it means to call add_note() and see the notes in the traceback.   If the ``__note__`` is not a sequence it is ignored, and any item in it which is not a string is skipped. So if you mess with it directly it shouldn't blow up.